### DISCUSSION TITLE
100 Node Limit

### DISCUSSION THREAD

**Discussion by c-heat16** [C-Accel-CRIPT/Python-SDK#391] [Development Discussions]
A dataset I am trying to upload to CRIPT contains 60+ different input materials, 130+ experiments, and will have 100s of products. When I try to upload the dataset to CRIPT, with a restricted number of output products, I receive the error message `cript.api.exceptions.CRIPTAPISaveError: API responded to POST with 'http:413 Number of nodes exceeds the amount allowed(100)' ... `.

Is there a way to increase this node limit? Is there a workaround? 

The script works if I only try to upload a single experiment.

**Comment by nh916** [comment]
hi @c-heat16, I have made a ticket for this, I am going to ask for the API to increase the limit to something higher for now and I can see how I can chunk the nodes to be under 100 nodes for future SDK versions

**Reply by c-heat16** [reply]
Thank you, @nh916 !

**Comment by c-heat16** [comment]
Its dawning on me that in the case of 100s of products for an experiment, the 'Graph preview' could get pretty unruly. I presume all of them likely wouldn't be shown. Would it perhaps be useful for products to have an 'appear_in_graph' property to allow users some control over which products get presented in case of many products? Some 'preferred' products will appear in the graph, but all products will be present in the panel on the right? Just a thought

**Reply by nh916** [reply]
That's a great idea, @c-heat16! We're definitely inclined to give scientists more control over data representation. Currently, our focus is ensuring system reliability, adding essential features, and enhancing the user experience. Once these are in place, refining data representation will be a top priority. I'll log this suggestion in our [SDK GitHub issue](https://github.com/C-Accel-CRIPT/Python-SDK/issues/395) to ensure we revisit it. Thank you so much for the suggestion!

**Comment by c-heat16** [comment]
While the node limit exists, is there a way to count the number of nodes currently associated with the project? 

IMO, it would be nice if we could check this value while in the process of creating the project - check in advance if adding additional materials/experiments/properties/what have you would go above the limit instead of 'guessing and checking', only getting a node count error after the entire project was created and attempted to be saved, removing aspects of the project until below the node limit.

**Comment by nh916** [comment]
> While the node limit exists, is there a way to count the number of nodes currently associated with the project?
> 
> IMO, it would be nice if we could check this value while in the process of creating the project - check in advance if adding additional materials/experiments/properties/what have you would go above the limit instead of 'guessing and checking', only getting a node count error after the entire project was created and attempted to be saved, removing aspects of the project until below the node limit.

Thank you for bringing this up. Currently, the SDK does not have a feature to automatically count the nodes within a project. I have thought about maybe overriding the len function within the base node, so calling `len()` on any node would show how many immediate child nodes are inside of it, for example `len(collection)` can show how many experiments are inside of it. However, I don't think this would be the full solution to what you're looking for. 

I am not currently sure how to write the code to get the total amounts of nodes and sub-objects within the project node, I'll have to really work through that. The code would probably be a DFS/BFS on either the project Python object or through the JSON and keep count of every object that it finds. I'll take notes on this feature and see how I can tackle it and incorporate it into future releases.

Currently, our API software engineer is planning to increase the node limit significantly from 100 to 1,000 soon. This should alleviate some of the current constraints, but more than 1,000 nodes will still have issues, but we might bump into it much less since the limit is much larger.
